### PR TITLE
Options

### PR DESCRIPTION
**Pull request by Highlife1911**

### CODE REVIEW

**Comment by Highlife1911** [comment]
This can be merged as is but there are 2 more ideas:
- [ ] shared_ptr in options should be replaced with unique_ptr.
- [ ] we should check for static option_map<Task> (void) functions.

File: source/tests/yats-test/options_test.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+{
+    static options_map<task> options()
+    {
+        return
+        {
+            { "value", &task::val }
+        };
+    }
+
+    output_bundle<slot<int, 0>> run(slot<int, 0> input)
```
Comment (lawben) [inline source/tests/yats-test/options_test.cpp]: do we really need the `output_bundle` if we return only one value? I have tests where this is not needed.

File: source/tests/yats-test/options_test.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        return std::make_tuple(static_cast<int>(input * val));
```
Comment (lawben) [inline source/tests/yats-test/options_test.cpp]: why is a `static_cast` necessary here?

File: source/yats/include/yats/options.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+class option_storage
+{
+public:
+    option_storage(options_map<Task> option)
+        : m_values(std::move(option))
+    {
+    }
+
+    template <typename Type>
+    void update(const std::string &key, Type value)
```
Comment (lawben) [inline source/yats/include/yats/options.h]: `std::string& key`

File: source/yats/include/yats/options.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        auto typed_member_pointer = static_cast<typed_member<Task, Type>*>(member)->pointer;
+        m_changes.push_back([typed_member_pointer, value = std::move(value)](Task *task) mutable
+        {
+            task->*typed_member_pointer = std::move(value);
+        });
+    }
+
+    void make_updates_visible(Task *task)
+    {
+        for (auto &change : m_changes)
```
Comment (lawben) [inline source/yats/include/yats/options.h]: could it happen that this never terminates? if we always update a value and thus `m_changes` is never empty

Comment (Highlife1911) [reply source/yats/include/yats/options.h]: this is not possible, but this part till contains a race condition.

**Pull request merged**